### PR TITLE
fix #5381: HDF5 Plotting Refactor

### LINKED ISSUE #5381: 20230301 Common H5 plotting prep
This is similar to #5380 but deals with hdf5 files. Apps which plot from hdf5 files include (opal, warppba, warpvnd, silas, omega). After #5380 is completed, a new class HDF5Util could be created which supports line and heat plots. This could possibly be expanded to include other common hdf5 file manipulation.

### PR DESCRIPTION
**Pull request by rorour**

### CODE REVIEW

File: sirepo/template/hdf5_util.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+from pykern.pkcollections import PKDict
+from pykern.pkdebug import pkdp
+from sirepo.template import template_common
+import numpy as np
+
+
+class HDF5Util:
+    # TODO(rorour): add docs for heatmap & lineplot
+    def __init__(self):
+        pass
```
Comment (moellep) [inline sirepo/template/hdf5_util.py:16]: Here it should take the hdf5 file in the constructor, and then when plotting, the file should be opened with h5py and the file handled passed to the callbacks if needed.

File: sirepo/template/hdf5_util.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                continue
+            plots[f] = PKDict(
+                label=plot_attrs.model[f],
+                dim=f,
+                points=[],
+                name=name_and_field[0],
+                index=plot_attrs.index(name_and_field),
+            )
+        plot_attrs.format_plots(plots)
+        x = plots.get(x_field)
```
Comment (moellep) [inline sirepo/template/hdf5_util.py:60]: here format_plots() could be passed the open file handle as an argument, so the caller doesn't need to open the hdf5 file beforehand.

**Review by moellep — CHANGES_REQUESTED** [review]

File: sirepo/template/hdf5_util.py (hunk 2 above)
Comment (rorour) [reply sirepo/template/hdf5_util.py:78]: I'm not totally understanding your vision here- let me know if this still needs to change

File: sirepo/template/hdf5_util.py (hunk 1 above)
Comment (rorour) [reply sirepo/template/hdf5_util.py]: d27e19ab27a41125f8de5d4f719ec9b485904168

File: sirepo/template/hdf5_util.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        x_field = "x"
+        y_field = "y"
+        plots = PKDict()
+        for f in (x_field, y_field):
+            n = plot_attrs.model[f]
+            plots[f] = PKDict(
+                name=n,
+                points=np.array(
+                    plot_attrs.file["/Step#{}/{}".format(plot_attrs.frame_index, n)]
+                ),
```
[the remark this answers is not in the document]
Comment (rorour) [reply sirepo/template/hdf5_util.py]: d27e19ab27a41125f8de5d4f719ec9b485904168

**Comment by rorour** [comment]
Do you know why this is failing ci? I saw this line in the log:
`import h5py found in: sirepo/template/hdf5_util.py`
would that be causing a failure?

For warppba extract_particle_report, I refactored to use template_common, but I wasn't sure about converting to use HDF5Util. What do you think is the best way to approach that?

**Comment by moellep** [comment]
The "import h5py found" can be fixed by adding hdf5_util to the exception list in test.sh:
https://github.com/radiasoft/sirepo/blob/master/test.sh#L27

That check was added due to problems loading apps if h5py wasn't imported in server.py
original bug: https://github.com/radiasoft/sirepo/issues/1151

In this case, we aren't adding apps, but moving the "import h5py" into the utility class, so it shouldn't be a problem.

**Comment by rorour** [comment]
@moellep Thanks, tests are passing now

**Review by moellep — APPROVED** [review]

**Pull request merged**